---
title: "Complementary analysis"
author: "J. Ignacio Lucas Lled√≥"
date: "25/11/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction
In the current version of the manuscript, the relationship between gut microbiota
composition and life history traits is explored with only a redundancy analysis
with a particular choice of candidate ASVs. The results are suggestive, but probably
not very robust to alternative choices of FDR and skewness thresholds, among other
more or less arbitrary choices (for example, the logarithmic transformation of relative
ASV abundances). Readers would probably thank an alternative approach, at least to be
convinced that a relationship exists between microbiota composition and male life
history traits. And the fact is that we did run a different analysis before the RDA,
which was based on a PCA of life history traits and a rotated MDS of microbiota
composition. Those were actually the results reported by Zahida in her dissertation,
and the main reason to exclude them from the manuscript was to simplify the exposition
of methods and results.

Here, I want to replicate the PCA and associated analyses, as well as compensate the
excessive attention given to ASVs. 

## Import data

```{r abundance}
load('../2020-04-22/Diversity.RData')
```

The file `.../2020-04-22/Diversity.RData` includes 16 objects, of which:

* **Ar** is the matrix of abundance data (in absolute counts) of 2659 ASVs (columns) in 177 samples.
* **Ar.prop** is the corresponding matrix of relative abundances (relative to each sample total).
* **SampleData** is a data frame with age, isoline, replicate and sequencing run of every sample.
* **Traits** is a data frame with taxonomy and abundance information on all 2659 ASVs.

The rest can be ignored. 

```{r ignoring}
rm(list=ls()[! ls() %in% c('Ar','Ar.prop','SampleData','Traits')])
```

Life history traits are stored in `../../data/LifeHistoryVar.txt`.

```{r LH}
LH <- read.table('../../data/LifeHistoryVar.txt', header=TRUE, row.names=1)
LH$CSrel <- LH$CSSlope / LH$CSearly
```

The `LH` data frame includes variables that will not be used and a couple of isolines
without microbiome data. Before cleaning it up, however, I run the multiple imputations,
in order to use all of the available information when imputing.

```{r LH_clean, message=FALSE, warning=FALSE}
library(mice)  # for imputation
# Centering. Using "apply()" instead would produce a matrix.
LH.centered <- as.data.frame(lapply(LH, function(x) (x - mean(x, na.rm=TRUE)) / sd(x, na.rm=TRUE)))
row.names(LH.centered) <- row.names(LH)
LH.imputation <- mice(LH.centered, m = 10, seed = 108, printFlag = FALSE)
LH.imputed <- complete(LH.imputation, action = 'all')
validIsolines  <- levels(SampleData$isoline)
validVariables <- c('AvLF','CSearly', 'CSSlope', 'EarlyRS', 'Rsen', 'ActuarialB')
LH.imputed <- lapply(LH.imputed, function(x, f1=validIsolines, f2=validVariables) x[f1, f2])
```

```{r reduced}
LH.reduced <- LH[! is.na(LH$ActuarialB), validVariables]
pairs(LH.reduced)
```

## PCA of *reduced* life history traits
The *reduced* data set excludes the two isolines for which the actuarial ageing
($\beta$ parameter of the Gompertz fit) could not be estimated, namely
`r paste(row.names(LH[is.na(LH$ActuarialB),]), sep=' and ')`. There are, thus,
27 isolines in this matrix, and six life history variables:

* Two functional measures:
  - **Early climbing speed**: How quickly males climb vertically when young, a proxy
    for their anti-predatory performance and general flying/reaction speed.
  - **Functional ageing**: Climbing speed slope, or the slope of the regression
    climbing speed $\sim$ age. This is a measure of how steeply climbing speed
    decreases with age, so a measure of functional senescence.
* Reproductive fitness:
  - **Early reproductive fitness**: Relative success of young males when
    competing against other males for reproduction.
  - **Reproductive senescence**: The difference in reproductive fitness between
    young and old males of each isoline, which gives us a measure of
    reproductive ageing.
* Actuarial ageing:
  - **Average lifespan** of the isoline.
  - **Acceleration of mortality rate**: This is the beta component in a Gompertz fit
    on the survival curve of each isoline. It gives us a measure of how mortality
    accelerates with age in each isoline. In combination with lifespan, it pretty
    much describes actuarial ageing.
    
```{r pca, fig.width=10, fig.height=10}
LH.pca <- prcomp(LH.reduced, center = TRUE, scale. = TRUE)
summary(LH.pca)
par(mfrow=c(2,2))
biplot(LH.pca, choices = c(1,2), xlim = c(-0.4, 0.4), ylim = c(-0.4, 0.4))
biplot(LH.pca, choices = c(3,2), xlim = c(-0.4, 0.4), ylim = c(-0.4, 0.4))
biplot(LH.pca, choices = c(1,3), xlim = c(-0.4, 0.4), ylim = c(0.4, -0.4))
par(mfrow=c(1,1))
```

This analysis suggests PC1 is dominated by variation in climbing speed, which
represents physical condition or performance. Note, however, that the two variables
`CSearly` and `CSSlope` are negatively correlated due to their mathematical dependence:
`CSSlope = CSlate - CSearly`. PC2 is mostly explained by early reproductive success
(`EarlyRS`). And PC3 is mostly explained by reproductive senescence and life span, with
positive values indicating faster reproductive senescence and shorter life span. The
three principal components found have clear biological interpretations, under this
selection of of life history variables.

I would like to consider an alternative selection of variables. Instead of `CSSlope`,
I would choose `CSlate`, which is not mathematically dependent on `CSearly`. And I would
log-transform `ActuarialB`, to correct its skewness and because I think it can make
biological sense. 

```{r PCA2, fig.width=10, fig.height=10}
LH.reduced2 <- LH[! is.na(LH$ActuarialB),
                  c('AvLF','CSearly', 'CSlate', 'EarlyRS', 'Rsen', 'ActuarialB')]
pairs(LH.reduced2)
LH.reduced2$ActuarialB <- log(LH.reduced2$ActuarialB)
LH.pca2 <- prcomp(LH.reduced2, center = TRUE, scale. = TRUE)
summary(LH.pca2)
par(mfrow=c(2,2))
biplot(LH.pca2, choices = c(1,2), xlim = c(-0.4, 0.4), ylim = c(-0.4, 0.4))
biplot(LH.pca2, choices = c(3,2), xlim = c(-0.4, 0.4), ylim = c(-0.4, 0.4))
biplot(LH.pca2, choices = c(1,3), xlim = c(-0.4, 0.4), ylim = c(0.4, -0.4))
par(mfrow=c(1,1))
```

The interpretation of these principal components is less straightforward. The fact that
most variables increase towards positive values of PC1 (except for reproductive senescence),
seems to reflect the fact that most variation in life history traits is due to genetic
quality differences among isolines, rather than to evolved strategies. However, `ActuarialB`
also loads positively on PC1, which contradicts the idea that positive PC1 simply means
better overall performance.

## Summary of abundance data
On `2020-02-27` I used the `phyloseq` package to merge abundance, taxonomy and
phylogenetic information in one object. After experiencing with some abundance
and prevalence thresholds for filtering ASVs, and also after exploring different
distance measures for ordination, I concluded that early and old samples differences
in taxonomic composition were rather mild. In particular, I used the quite permissive
minimum average proportion of $10^{-5}$. Also, the lack of familiarity with `phyloseq`
makes me doubt the data was handled well. Most striking is the fact that presence-absence
measures of composition distance among samples were the only ones that produced
ordinations that clearly separated early from late samples.

Such a use of low-abundance ASVs and presence-absence measures of composition
distance are very much at odds with the current understanding of the data set.
According to our analysis in `2020-04-27`, not only phylogenetic information
but also abundance should be included in the distance measures in order to
appreciate a pattern that is both quite apparent at the genus level and very
much reasonable: the loss of $\beta$ diversity with age. 

Thus, instead of replicating the original characterization of microbiome composition
from `2020-02-27`, I will start over. First, before filtering by abundance, I will
generate the matrix of genus abundance.

### Genus abundance

```{r genera, warning=FALSE}
library(ggplot2)
stopifnot(all.equal(rownames(Traits), colnames(Ar)))
Traits[is.na(Traits$Genus), 'Genus'] <- 'unknown'
# Having no NA values in Genus, I can use "==" below, instead of "%in%". 
# Using sapply(..., simplify=FALSE) instead of lapply() I get columns named by genus
Genera.counts <- do.call(cbind,
                         sapply(unique(Traits$Genus),
                                function(x, G=Traits, M=Ar) {
                                  f <- rownames(G[G$Genus == x,])
                                  rowSums(M[, f, drop=FALSE])
                                },
                                simplify = FALSE))
# I generate genus frequencies by adding up ASV frequencies, instead of normalizing
# the counts in case ASV relative frequencies do not add up to 1, if some ASVs
# have already been filtered out.
Genera.prop <- do.call(cbind,
                       sapply(unique(Traits$Genus),
                              function(x, G=Traits, M=Ar.prop) {
                                f <- rownames(G[G$Genus == x,])
                                rowSums(M[, f, drop = FALSE])
                              },
                              simplify = FALSE))

Taxonomy <- unique(Traits[Traits$Genus != 'unknown', c('Genus','Family','Order','Class','Phylum')])
Taxonomy <- rbind(Taxonomy, rep('unknown', 5))
row.names(Taxonomy) <- Taxonomy$Genus

stopifnot(all.equal(row.names(SampleData), row.names(Genera.counts)))
stopifnot(all.equal(row.names(SampleData), row.names(Genera.prop)))
Genera.summary <- data.frame(prevalence = colSums(Genera.counts > 0),
                             meanProp   = colSums(Genera.prop) / dim(Genera.prop)[1],
                             EarlyPrev  = colSums(Genera.counts[SampleData$age == 'Early',] > 0),
                             LatePrev   = colSums(Genera.counts[SampleData$age == 'Late', ] > 0),
                             EarlyMeanProp = colSums(Genera.prop[SampleData$age == 'Early',]) /
                               sum(SampleData$age == 'Early'),
                             LateMeanProp  = colSums(Genera.prop[SampleData$age == 'Late', ]) /
                               sum(SampleData$age == 'Late'),
                             Family = Taxonomy[colnames(Genera.prop),'Family'],
                             Order =  Taxonomy[colnames(Genera.prop),'Order'],
                             Class =  Taxonomy[colnames(Genera.prop),'Class'],
                             Phylum = Taxonomy[colnames(Genera.prop),'Phylum'])
ggplot(data=Genera.summary, mapping=aes(x=meanProp, y=prevalence, color=Phylum)) +
  geom_point() + scale_x_log10()
ggplot() + geom_abline(slope=1, intercept=0) +
  geom_point(data=Genera.summary, mapping=aes(x=EarlyMeanProp, y=LateMeanProp, color=Phylum)) +
  scale_x_log10() + scale_y_log10()
```

### ASV abundance

```{r change, warning=FALSE}
Traits$AbundantGenus <- Traits$Genus
RareGenera <- unique(row.names(Genera.summary[Genera.summary$meanProp < 0.001,]))
Traits[Traits$Genus %in% RareGenera, 'AbundantGenus'] <- 'Others'
ggplot() + geom_abline(slope=1, intercept=0) +
  geom_point(data=Traits, mapping=aes(x=EA.meanProp, y=LA.meanProp, color=AbundantGenus), alpha=0.5) +
  scale_x_log10() + scale_y_log10() + xlab('Average early relative abundance') +
  ylab('Average late relative abundance') + ggtitle('ASV average relative abundances')
```

The plot above represents the relationship between early and late
ASV average relative abundances, in logarithmic scale. The black
line in the background corresponds to no change in average abundance
between early and late samples. Amplicons below the black line
reduced average relative abundance, and those above, increased it.
The significance of the change cannot be assessed, because the biological
variation of an amplicon's abundance across replicates is not shown.

Color indicates the genus, if it is an abundant genus.

What is remarkable in that plot is that several ASVs of high average
relative abundance cluster in lines of almost unit slope. Because it's
in logarithmic scale, those clusters respresent subsets of ASVs with
an almost identical fold change of average relative abundances.

ASVs colored as "Others" correspond to several rare genera, most of which
becoming even rarer with age. A similar behaviour is observed in
*Methylobacterium*: while as a set, the genus is quite abundant,
few of its ASVs ever reach a relative abundance of $10^{-4}$, and
they all reduce their representation with age.

It is also remarkable that clusters are composed of ASVs from the
same genus. We can see that *Acetobacter* includes three kinds of
ASVs: the very rare ones near the bottom-left corner, the ones that
keep an almost constant abundance, those that increase and those that
decrease in abundace.

The figure reveals a problem with the taxonomic attribution: we lack
resolution in the *Acetobacter* genus, where different subsets behave
in consistent and different ways. They probably represent different
lineages or species. Without being able to refer to those categories,
the description of sample' composition will be incomplete and
therefore less likely to correlate with life history trait variation.

At least, the figure offers also some guidance to filtering by
abundance.