---
title: "Phyloseq analysis"
author: "J. Ignacio Lucas Lled√≥"
date: "27/2/2020"
output: html_document
params:
  DATA: '../2020-02-11/taxonomy.RData'
  RUN1: '../../data/RUNGEN52_2019/fastq'
  RUN2: '../../data/RUNGEN66/fastq'
bibliography:
  phyloseq.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r setup2, message=FALSE}
library(phyloseq)
library(ggplot2)
library(gridExtra)
library(Biostrings)
library(DECIPHER)
library(phangorn)
library(plyr)
load(params$DATA)
NUM_THREADS=10
ls()
```

# Naming sequencies

Above I loaded the taxonomy data, including the table of abundances (`SeqTabNoChim`)
and the assigned taxonomy (`taxa`). I name all sequences with a taxonomy attribution.

```{r naming}
dna <- DNAStringSet(row.names(taxa))
names(dna) <- sprintf("A%04i", seq(dim(taxa)[1]))
row.names(taxa)        <- names(dna)[match(row.names(taxa), dna)] 
colnames(SeqTabNoChim) <- names(dna)[match(colnames(SeqTabNoChim), dna)]
```

# Sample data

```{r sampleData}
SampleData <- data.frame(
  age = factor(substr(rownames(SeqTabNoChim), 1, 1), levels=c('E','L'), labels=c('Early','Late')),
  isoline = factor(stringr::str_extract(rownames(SeqTabNoChim), "[[:digit:]]+"),
                   levels=as.character(c(6,10,11,12,14,15,17,19,20,22,23,24,25,26,27,28,29,30,31,33,35,36,38,39)),
                   ordered=TRUE),
  replicate = gsub("([EL][[:digit:]]+|.fastq.gz)", "", rownames(SeqTabNoChim)),
  seqrun = factor(1, levels=c(1,2))
)
row.names(SampleData) <- rownames(SeqTabNoChim)
run2 <- paste(substring(SampleData$age, 1, 1),
              SampleData$isoline,
              SampleData$replicate,
              '_R1.fastq.gz', sep='') %in% dir(params$RUN2)
SampleData[run2, 'seqrun'] <- 2
rm(run2)
```

# Additional filtering
Recall the `SeqTabNoChim` matrix of abundances has been filtered before: only
sequence lengths in the range `r range(nchar(colnames(SeqTabNoChim)))` are
present, and sequences of unknown or unreliable Phylum have been removed
(though they are still present in `taxa`). Following [@Callahan2016],
I take a look at *prevalence* (number of samples a taxon is observed at
least once), and its relationship with total abundance.

```{r prevalence, warning=FALSE, message=FALSE}
Prevalence <- data.frame(
  Prev.Early = colSums(SeqTabNoChim[SampleData$age %in% 'Early',] > 0),
  Prev.Late  = colSums(SeqTabNoChim[SampleData$age %in% 'Late',] > 0),
  Abun.Early = colSums(SeqTabNoChim[SampleData$age %in% 'Early',]),
  Abun.Late  = colSums(SeqTabNoChim[SampleData$age %in% 'Late',]),
  prevalence = colSums(SeqTabNoChim > 0),   # taxa are columns
  abundance  = colSums(SeqTabNoChim),
  Phylum     = factor(taxa[colnames(SeqTabNoChim), 'Phylum']),
  Order      = factor(taxa[colnames(SeqTabNoChim), 'Order'])
)
Proportions <- apply(SeqTabNoChim, 1, function(x) x/sum(x))
Prevalence$meanProp <- apply(Proportions, 1, mean)
Prevalence$meanPropEarly <- apply(Proportions[, SampleData$age %in% 'Early'], 1, mean)
Prevalence$meanPropLate  <- apply(Proportions[, SampleData$age %in% 'Late'],  1, mean)
ggplot(Prevalence, aes(x=meanProp, y=prevalence)) +
  geom_point(size=0.5) + scale_x_log10() + facet_wrap(~Phylum)
#ggplot(Prevalence[Prevalence$Phylum=='Proteobacteria',], aes(x=meanProp, y=prevalence, color=Order)) +
#  geom_point() + scale_x_log10()
ggplot(Prevalence[Prevalence$Phylum=='Proteobacteria',], aes(x=meanPropEarly, y=Prev.Early, color=Order)) +
  geom_point() + scale_x_log10() + ggtitle('Proteobacteria in early samples')
ggplot(Prevalence[Prevalence$Phylum=='Proteobacteria',], aes(x=meanPropLate, y=Prev.Late, color=Order)) +
  geom_point() + scale_x_log10() + ggtitle('Proteobacteria in late samples')
ggplot(Prevalence, aes(x=meanPropEarly, y=meanPropLate)) + geom_point() +
  scale_x_log10() + scale_y_log10() + geom_abline(slope=1, intercept=0, color='red')
```

In the last plot, I see among the most abundant amplicons, many that keep the same
relative abundance in early and late times, some that are more abundant late than
early, and some that are more abundant early.

In the [preprocessing tutorial](https://joey711.github.io/phyloseq/preprocess.html)
some filters are suggested that can be applied to a `phyloseq` object. Here,
I filter sequences before creating the phyloseq object, because that way the
alignment and phylogenetic tree reconstruction steps will be faster.

[@Callahan2016] suggests to filter by prevalence. They remove taxa that are not
present in at least 5% of samples. However, the [preprocessing tutorial](https://joey711.github.io/phyloseq/preprocess.html)
suggests to filter by mean relative abundance, using $10^{-5}$ as the threshold.

On a first run, I imposed the thresholds on global measures of mean proportion and
prevalence. However, I realize that early and late samples should be allowed to
have different compositions. I use now a filter that respects the independence of
early and late samples.

```{r filter}
filter <- (Prevalence$meanPropEarly >= 1.0e-05 | Prevalence$meanPropLate >= 1.0e-05) & 
  (Prevalence$Prev.Early > 10 | Prevalence$Prev.Late > 10)
AmpFilt <- sort(colnames(SeqTabNoChim)[filter])
SeqTabClean <- SeqTabNoChim[,filter]
dim(SeqTabClean)
```

# Alignment and tree
I had aligned them before, on `2019-12-12`. But for the
sake of clarity, I align them here as well, in order to create a phylogenetic
tree and include that information in the `phyloseq` object. I align only the
filtered sequences.

Below, I use conditional execution as a *home made* substitute for the
`cache=TRUE` option, which does not seem to work well sometimes.

```{r alignment, message=FALSE, warning=FALSE}
if (file.exists('alignment.RData')) {
   load('alignment.RData')
} else {
   alignment <- AlignSeqs(RNAStringSet(dna[AmpFilt]),
                          iterations=10, refinements=10, processors=NUM_THREADS,
                          verbose=FALSE)
   BrowseSeqs(alignment, htmlFile='alignment.html', openURL=FALSE)
   save(alignment, file='alignment.RData')
   # Forces removal of tree upon update of alignment:
   if (file.exists('fitGTR.RData')) file.remove('fitGTR.RData')
}
```

Next few lines mostly copied from [@Callahan2016]. I suppose fitting the
tree by maximum likelihood may take a while. This is one of the chunks where
I would use the `cache=TRUE` option to prevent it from running if the
results are saved (and neither the input nor the code changed since the
time the results were saved). However, the chunk's cache does not seem
to work properly. It may be an RStudio issue. I resort to save the results
explicitly, and make the chunk's execution conditional on their absence.
I will have to remove the results manually if I want them updated.

```{r tree, message=FALSE}
if (file.exists('fitGTR.RData')) {
   load('fitGTR.RData')
} else {
   phang.align <- phyDat(as(DNAStringSet(alignment), "matrix"), type="DNA")
   dm <- dist.ml(phang.align)
   treeNJ <- NJ(dm) # Note, tip order != sequence order
   fit = pml(treeNJ, data=phang.align)

   fitGTR <- update(fit, k=4, inv=0.2)
   fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                      rearrangement = "stochastic", control = pml.control(trace = 0))
   detach("package:phangorn", unload=TRUE)   # avoids name clashes.
   save(fitGTR, file='fitGTR.RData')
}
```

# Phyloseq object
Now, I start using the `phyloseq` package [@McMurdie2013].
Row names in the matrix of abundances `SeqTabClean` contains the fastq file names
corresponding to each library, where the three variables identifying the samples are
encoded, namely the sampling time (or fly *age*), the isoline, and the replicate.

```{r data}
SampleData$simpson <- vegan::diversity(SeqTabClean, index='simpson', MARGIN=1)
SampleData$shannon <- vegan::diversity(SeqTabClean, index='shannon', MARGIN=1)

ps <- phyloseq(otu_table(SeqTabClean, taxa_are_rows=FALSE), 
               sample_data(SampleData), 
               tax_table(taxa[AmpFilt,]),
               phy_tree(fitGTR$tree))
ps <- merge_phyloseq(ps, dna[AmpFilt])
ps
plot_richness(ps, x="isoline", measures=c("Shannon", "Simpson"), color="age", shape="seqrun")
```

To generate a `phyloseq` object with relative abundance (proportions),
I prefer to create it *de novo* from the `PropClean` matrix created
before, instead of using the `transform_sample_counts` function. The
reason is that I want to use the original, true values of relative
abundance, and not the proportions calculated on a subset of the
original taxa.

```{r distance, message=FALSE, warning=FALSE}
ps.prop <- phyloseq(otu_table(Proportions[AmpFilt,], taxa_are_rows=TRUE),
                    sample_data(SampleData),
                    tax_table(taxa[AmpFilt,]),
                    phy_tree(fitGTR$tree))

save(ps, ps.prop, file='ps.RData')

ggplot(psmelt(ps.prop), aes(x=age, y=Abundance)) +
  geom_violin() + facet_wrap(~Phylum) + scale_y_log10()

ps.prot <- subset_taxa(ps.prop, Phylum %in% c("Proteobacteria"))
ggplot(psmelt(ps.prot), aes(x=age, y=Abundance)) +
  geom_violin() + facet_wrap(~Order) + scale_y_log10()

ord.nmds.bray <- ordinate(ps.prop, method="NMDS", distance="bray", trymax=1000, trace=0)
# The "trace=0" above supresses the long output.
plot_ordination(ps.prop, ord.nmds.bray, color="age", title="Bray NMDS")
```

# Reduced dataset

As mentioned before, a more aggressive filtering can be applied to
remove taxa that have not been observed a minimum number of times (4)
in a minimum number of samples (20%). I show the code below, but
don't run it now. Previous runs suggest that reducing the dataset so
much removes the main differences between early and late communities.

```{r CV, eval=FALSE}
ps.clean2 <- filter_taxa(ps, function(x) sum(x > 3) > (0.2*length(x)), TRUE)
total <- median(sample_sums(ps.clean2))
standf <- function(x, t=total) round(t * (x / sum(x)))
ps.clean2 <- transform_sample_counts(ps.clean2, standf)
# Now abundances are standardized to median sequencing depth.
ps.clean2
# In phyloseq tutorial they reverse the "<" operator, effectively keeping only
# taxa with high CV, for plotting purposes. In any case, the filter below does
# not remove much, after the previous steps. 
ps.clean2 <- filter_taxa(ps.clean2, function(x) sd(x)/mean(x) <= 3.5, TRUE)
ps.clean2
table(ps.clean2@tax_table@.Data[,'Phylum'])
table(ps.clean2@tax_table@.Data[,'Genus'])
```

# Ordination
Following the [tutorial](http://joey711.github.io/phyloseq/distance.html) on distances.
Note that here I use MDS, while the plot produced above was done using NMDS.

```{r distance1, fig.width=15, fig.height=15}
dist_methods <- unlist(distanceMethodList)
# in addition to designdist, I remove those that cause errors:
dist_methods <- dist_methods[! names(dist_methods) %in% c('vegdist1', 'designdist', 'vegdist2', 'dist3')]
plist <- vector("list", length(dist_methods))
names(plist) <- dist_methods
# Abundance should be integer, but standardized.
total <- median(sample_sums(ps))
standf <- function(x, t=total) round(t * (x / sum(x)))
ps.used <- transform_sample_counts(ps, standf, t=median(sample_sums(ps)))
#ps.used <- ps.clean2
for( i in dist_methods ){
    # Calculate distance matrix
    iDist <- phyloseq::distance(ps.used, method=i)
    # Calculate ordination
    iMDS  <- ordinate(ps.used, "MDS", distance=iDist)
    ## Make plot
    # Don't carry over previous plot (if error, p will be blank)
    p <- NULL
    # Create plot, store as temp variable, p
    p <- plot_ordination(ps.used, iMDS, color="age")
    # Add title to each plot
    p <- p + ggtitle(paste("MDS using distance method ", i, sep=""))
    # Save the graphic to file.
    plist[[i]] = p
}
df = ldply(plist, function(x) x$data)
names(df)[1] <- "Distance"
p <- ggplot(df, aes(Axis.1, Axis.2, color=age))
p <- p + geom_point(size=3, alpha=0.5)
p <- p + facet_wrap(~Distance, scales="free")
p <- p + ggtitle("MDS on various distance metrics")
p
```

In the plot above, the ordinations that separate better between old and young
microbiomes seem to be those that use one of the distance methods implemented
in the `betadiver()` function of the `vegan` package, such as *-1*, *-2*, *-3*,
*19*, *cc*, *e*, *g*, *hk*... `betadiver()` calculates either $\Beta$ diversity
indices or *distances* among local paterns of presence-absence of species.
The binary distance, from the `dist()` function also works well, and it also
treats abundances as presence-absence, actually.

When using `ps.prop`, with `r dim(ps.prop@otu_table@.Data)[1]` taxa, samples are
much better separate in early and late ages than if using the reduced `ps.clean2`
dataset.

Taken together, these results suggest that differences between young and old
microbiomes are subtle and can only be appreciated when looking at the presence
or absence of both low and high frequency taxa. To put it another way, they do
not differ much in the most abundant taxa.

The same is true for the batch effect. The same distance methods that separate
young and old microbiomes also distinguish quite well between the two sequencing
runs. Curiously, the discrimination between sequencing runs is almos orthogonal
to that of young and old samples. That is, the two first axes of ordinations
based on presence-absence distances correspond to the two main drivers of diversity
among our samples: age and batch effect. Axis 1 mostly corresponds with the
age of the samples, and Axis 2, with the sequencing batch. Although a small
counter-clock wise rotation would help the interpretation of the axes.

Just to illustrate the point:

```{r binary}
p <- ggplot(df[df$Distance == 'binary',], aes(Axis.1, Axis.2, color=age))
p <- p + geom_point(size=3, alpha=0.5)
p <- p + ggtitle("MDS on a binary distance. Coloring by age")
p
p <- ggplot(df[df$Distance == 'binary',], aes(Axis.1, Axis.2, color=seqrun))
p <- p + geom_point(size=3, alpha=0.5)
p <- p + ggtitle("MDS on a binary distance. Coloring by sequencing run")
p
```

# Rotation

I found that the `recluster` package can rotate the coordinates of an MDS to make
a line be horizontal. Ideally, the line would be a the boundary decision of a linear
discriminant analysis. But I have not found a way to extract the linear expression
of the boundary from the LDA result. Below, I use a manual approximation.

```{r rotate}
library(recluster)
BinaryMDS <- df[df$Distance == 'binary',]
tilted <- recluster.rotate(BinaryMDS[,c('Axis.1','Axis.2')], m=-0.47, q=-0.027)
BinaryMDS$Rotated1 <- tilted$Axis.1
BinaryMDS$Rotated2 <- tilted$Axis.2
rm(tilted)
ggplot(BinaryMDS, aes(x=Rotated1, y=Rotated2, color=seqrun)) +
  geom_point(size=3, alpha=0.5) +
  ggtitle("Slightly rotated MDS on binary distance")
ggplot(BinaryMDS, aes(x=Rotated1, y=Rotated2, color=age)) +
  geom_point(size=3, alpha=0.5)
p <- ggplot(BinaryMDS, aes(x=Rotated1, y=Rotated2, color=age, shape=seqrun)) +
  geom_point(size=3, alpha=0.5) + xlab('Rotated axis 1') + ylab('Rotated axis 2') +
  ggtitle('Multidimensional scaling on binary distances')
ggsave(filename = 'MDS_rotated.png', plot=p)
```

# Acetobacter/Lactobacillus ratio

```{r ratio}
Lactobacillus <- AmpFilt[AmpFilt %in% row.names(taxa[taxa[,'Genus'] %in% 'Lactobacillus',])]
Acetobacter   <- AmpFilt[AmpFilt %in% row.names(taxa[taxa[,'Genus'] %in% 'Acetobacter',])]
LactoProp <- colSums(Proportions[Lactobacillus,])
AcetoProp <- colSums(Proportions[Acetobacter,])
summary(LactoProp)
summary(AcetoProp)
SampleData$LactoAcetoRatio <- LactoProp / AcetoProp
# It includes 0, but not Inf. 
ggplot(SampleData, aes(x=age, y=LactoAcetoRatio)) + 
  geom_boxplot(notch=TRUE) + scale_y_log10()
ggplot(SampleData, aes(x=seqrun, y=LactoAcetoRatio)) +
  geom_boxplot(notch=TRUE) + scale_y_log10()
```

# Export data

Here I produce a `csv` file with the `SampleData` dataframe, which includes
Simpson and Shannon's diversity indices and the rotated coordinates of the
MDS produced with the binary distance. The rotation is meant to make the MDS
axis more readily interpretable as age (axis 1) and batch (axis 2) effects.
I also include the original axis, just in case.

```{r sesionInfo}
# Actually, samples are in the same order in SampleData and BinaryMDS.
SampleData$RotAxis1 <- BinaryMDS$Rotated1
SampleData$RotAxis2 <- BinaryMDS$Rotated2
SampleData$OriAxis1 <- BinaryMDS$Axis.1
SampleData$OriAxis2 <- BinaryMDS$Axis.2
write.csv(SampleData, file='DiversityAbundance.csv', quote=FALSE)
sessionInfo()
```

# References
