\documentclass[a4paper,12pt]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{booktabs}
\usepackage{natbib}
\usepackage{amsmath}
\usepackage[colorlinks=true]{hyperref}
\usepackage[official]{eurosym}
\title{Genotyping by sequencing protocol}
\begin{document}
\maketitle
\section{DNA extraction}
The goal is to obtain at least 500 ng of high quality DNA from single individuals. 
\subsection{First session}
You need a DNeasy Blood \& Tissue Kit (Qiagen), and RNase A. Have some liquid nitrogen ready in a container, clean plastic pestles and an electric screw driver. Buffer ATL may form precipitate upon storage. If necessary, warm to 56$^\circ$C until precipitates fully dissolve.
\begin{enumerate}
\item Put one fly in a 1.5 ml tube and immediately, on ice. Add a small amount of autoclaved sea sand. Keep the tube open and deep the bottom of it in the liquid nitrogen to freeze the sample, with care not to freeze your fingertips. Use the screw driver with a pestle fitted on it to grind the fly. Press hard the pestle against the tube. Alternate between grinding and cooling the tube down in the liquid nitrogen.
\item When the fly is powdered, let the tube recover a temperature >0$^\circ$C. Add 180 $\mu$l of buffer ATL and remove the pestle with care of leaving the liquid in the tube. Then, add 20 $\mu$l of proteinase K, vortex, and incubate at 56$^\circ$C overnight on a rocking platform. Make sure to seal the tubes with parafilm to reduce evaporation.
\end{enumerate}
\subsection{Second session}
\begin{enumerate}
\item Just like the ATL buffer, buffer AL may also form precipitates. Warm it up if necessary to dissolve them. Make sure that buffers AW1 and AW2 have the ethanol added.
\item After the incubation, bring samples to room temperature. Centrifuge for 2 min at 8000 $\times g$ (10000 rpm?), at 20$^\circ$C. Transfer supernatant to a new tube. This removes the sand.
\item Add 4 $\mu$l RNase A (100 mg/ml), mix by vortexing, and incubate for 2 min at room temperature.
\item Vortex for 15 s. Add 200 $\mu$l buffer AL to the sample, and mix thoroughly by vortexing. Then add 200 $\mu$l ethanol (96-100\%), and mix again thoroughly by vortexing. It is essential that the sample, buffer AL, and ethanol are mixed immediately and thoroughly by vortexing or pipetting to yield a homogeneous solution. Buffer AL and ethanol can be premixed and added together in one step to save time when processing multiple samples.
\item Pipet the mixture from the previous step (including any precipitate) into the DNeasy Mini spin column placed in a 2 ml collection tube. Centrifuge at $\geq6000 \times g$ (8000 rpm) for 1 min. Discard flow-through and collection tube. 
\item Place the DNeasy Mini spin column in a new 2 ml collection tube, add 500 $\mu$l buffer AW1, and centrifuge for 1 min at $\geq6000 \times g$ (8000 rpm). Discard flow-through and collection tube.
\item Place the DNeasy Mini spin column in a new 2 ml collection tube, add 500 $\mu$l buffer AW2, and centrifuge for 3 min at 20000$\geq \times g$ (14000 rpm) to dry the DNeasy membrane. Discard flow-through and collection tube. It is important to dry the membrane, since residual ethanol may interfere with subsequent reactions. This centrifugation step ensures that no residual ethanol will be carried over during the following elution. Following the centrifugation step, remove the DNeasy Mini spin column carefully so that the column does not come into contact with the flow-through, since this will result in carryover ethanol. If carryover of ethanol occurs, empty the collection tube, then reuse it in another centrifugation for 1 min at 20000$\times g$ (14000 rpm).
\item Place the DNeasy Mini spin column in a clean 1.5 ml or 2 ml microcentrifuge tube (not provided in the DNeasy kit), and pipet 200 $\mu$l buffer AE directly onto the DNeasy membrane. Incubate at room temperature for 1 min, and then centrifuge for 1 min at $\geq6000 \times g$ (8000 rpm) to elute. Elution with 100 $\mu$l (instead of 200 $\mu$l) increases the final DNA concentration in the eluate, but also decreases the overall DNA yield. For maximum DNA yield, repeat elution once more. Do not elute more than 200 $\mu$l into a 1.5 ml microcentrifuge tube because the DNeasy Mini spin column will come into contact with the eluate.
\end{enumerate}

\section{Digestion}
I digest the DNA with only one enzymee. Table~\ref{tau:digest} shows the composition of a typical digestion reaction. The DNA volume varies among samples. When deciding the final volume, it is important to keep in mind the following advice:
\begin{enumerate}
   \item If DNA is in a buffer with some salt (such as TE), then the DNA volume should \textbf{not} be more than 25\% of the reaction volume. Otherwise, the salt in the buffer can inhibit the enzyme.
   \item If DNA is very diluted and the DNA volume is higher than 100 $\mu$l, it may be a good idea to concentrate the DNA first. However, a highly diluted DNA is apparently the last of the concerns in restriction reactions (mental note: I wonder if it can be compensated to some extent with longer incubation time).
   \item As a rule of thumb: use 1 $\mu$l of enzyme (10000 units/ml) per 1 $\mu$g of DNA.
   \item Enzymes are in 50\% glycerol (to avoid freezing). Always keep the total concentration of glycerol in the reaction below 5\%.
   \item DNA concentration in the reaction should be at 20-100 ng/$\mu$l.
\end{enumerate}

\begin{table}
\begin{center}
\caption{Typical restriction digest.}\label{tau:digest}
\vspace{0.3cm}
\begin{tabular}{ll}
\toprule
Restriction enzyme&10 units is sufficient, generaly 1$\mu$l is used.\\
DNA&$x \mu$l ($\sim1$ $\mu$g)\\
10$\times$ NEBuffer& $0.4x$ $\mu$l.\\
Total reaction volume&$4x$ $\mu$l.\\
H$_2$O&$2.6x - 1 \mu$l\\
Incubation time&1 hour (I don't think it hurts if more).\\
Incubation temperature&37$^{\circ}$C for most enzymes.\\
\bottomrule
\end{tabular}
\end{center}
\end{table}

Peterson et al. \cite{Peterson2012} recommend not to stop the digestion by heat-inactivation, to prevent a bias against shorter and AT-rich fragments, which would de-naturalize during heating and then loose the chance to ligate the adapters. Instead, we are supposed to use a Spin Column or magnetic beads (see below) to clean the reaction product and concentrate the DNA. Note, however, that similar protocols do not hesitate to use heat inactivation \cite{Andolfatto2011,Etter2011}.

Sibelle Vila√ßa (personal communication) recommended to run the digestion and the ligation in the same step. That is, all the enzymes were together in the tube, first giving them the chance to digest, and then to ligate, without any clean up step in between. She reported lower levels of chimeras when doing this. It may also improve the yield, since we save a clean up step. I understand now that to combine digestion and ligation in one reaction, it may be crucial to have adapters that do not reproduce the recognition site when ligated. This is possible as long as the overhang does not include the whole recognition site.

\subsection{Clean up with magnetic beads}
You can consider skipping this step (see above).
\begin{enumerate}
\item Prepare twice as much 80\% ethanol as total DNA volume you have among all the tubes, or a bit more. Prepare at least 400 $\mu$l of 80\% ethanol per sample. If ethanol is 80\%, it does not need to be freshly prepared \cite{Bronner2009}.
\item Vortex High Prep PCR beads thoroughly.
\item Add beads in a 1.2:1 ratio\footnote{1.5 ratio would allow binding of short fragments as well. Last time, I used a 1:1 ratio, and I noticed absence of fragments below 200 bp, and a likely depletion of fragments below 1000 bp.} to the restriction (or PCR) reaction and vortex well (30 s; or pipette mixing).
\item Incubate at room temperature for 5 min (or more; this is an important step) without shaking.
\item Place tube on magnetic rack for 3 min, or until the solution is clear (the time depends on the volume of the solution).
\item Remove supernatant carefully without disturbing the beads. Make sure you have NO beads in the pipet.
\item Add the same amount of freshly prepared ethanol as supernatant you removed (at least 200 $\mu$l), so that the alcohol covers all the beads.
\item Let stand for a minute.
\item Remove supernatant.
\item Add the same amount of 80\% ethanol again.
\item Let stand for another minute.
\item Remove the supernatant.
\item Air dry pellet 5 minutes at room temperature. You may see a crack in the pellet. If you over-dry the beads, you will see many cracks. If you under-dry the beads, the DNA recovery rate will be lower, due to the remaining ethanol. 
\item Add 1$\times$ TE buffer, or water.
\item Vortex for 30 s or pipette mix 10 times. The liquid level should be high enough to contact the magnetic beads. If not, extra vortexing is required, and may not be sufficient to fully elute all the DNA.
\item Incubate 2 min at room temperature.
\item Place tube on magnetic rack for 2 min, or until the solution is clear.
\item Transfer supernatant to a fresh tube.
\end{enumerate}

About the final volume to elute the DNA, I have been successful with very low ammounts (10-20 $\mu$l), if the pellet was not too dry. It's easier with more. After the digestion reaction, I recover the samples with the amount of TE that would give me the same concentration in all samples, in the order of 10 ng/$\mu$l, assuming that all the original amount of DNA is recovered.

\subsection{Estimate distribution of fragment sizes with Agilent BioAnalyzer}
This is a tricky machine. Before using it, it is convenient to clean the electrodes: just put 450 $\mu$l H$_2$O in a cleaning chip, place it in the machine (with lid closed), and let it sit for a little while. After removing the water chip, leave the lid of the machine open for a couple of minutes to dry de electrodes. The machine does not have to be turned on for this. You can rinse the chip with destilled water and re-use it later.

Before the assay, you should also set up the computer: turn it on, open the software, choose the machine, select the type of assay, and name the samples.

The assay chip has 11 spots available for samples. When loading it, it is recommended not to press the pippette to the second stop, to avoid the introduction of air bubbles. Loading should be accurate and not too slow. Once the chip is ready, it should be used immediately. Every assay costs on the order of \euro{}50.00.

Follow the instructions in the original protocol.

\section{Ligation}
\subsection{Prepare annealed adapters stock}
This is a combination of three protocols \cite{Andolfatto2011,Etter2011,Peterson2012}. Oligos are delivered dry, and need to be suspended in any desired volume or concentration. The delivery documentation suggests 100 pmol/$\mu$l (that is, 100 $\mu$M).

\begin{table}
\caption{Annealing buffer (100 $\mu$l, 10X).}\label{tau:buffer}
\vspace*{0.3cm}
\begin{tabular}{lrrr}
\toprule
Reagent&Amount to add&Final concentration\\
\midrule
Tris-HCl, pH 8, 1 M&10 $\mu$l&100 mM\\
EDTA, 0.5 M&2 $\mu$l&10 mM\\
NaCl (58.44 g/mol)&2.92 mg&500 mM\\
H$_2$O&88 $\mu$l&\\
\bottomrule
\end{tabular}
\end{table}

\begin{enumerate}
\item Prepare the annealing buffer stock (10X): 100 mM Tris HCl (pH 8), 500 mM NaCl, 1 mM EDTA (see table~\ref{tau:buffer}).
\item Suspend the oligos in the volumes required to have them in 100 $\mu$M (100 pmol/$\mu$l). Use TE buffer.
\item To produce 10 $\mu$l of annealed stock from each adapter, at 15 $\mu$M (15 pmol/$\mu$l), combine: 1.5 $\mu$l of the top oligo (100 $\mu$M), 1.5 $\mu$l of the bottom oligo (100 $\mu$M), 1 $\mu$l 10$\times$ annealing buffer (table~\ref{tau:buffer}), and 6 $\mu$l of H$_2$O. 
\item In a thermocycler, incubate at 97.5 $^{\circ}$C for 2.5 minutes, and then cool at a rate not greater than 3 $^{\circ}$C per minute, until the solution reaches a temperature of 21 $^{\circ}$C. Hold at 4 $^{\circ}$C.
\item The final working strength concentration of annealed adapters is calculated for each sample to have 7 times more adapters than fragment ends in the sample, in a volume of 1 or 2 $\mu$l. The required working concentration must be lower than the concentration of the annealed adapter stock (it can be lowered by increasing the volume added per reaction).
\end{enumerate}

\begin{table}
\caption{Example of volumes of ligation reactions. All volumes are in $\mu$l. T4 ligase is at 2000000 U/ml.}\label{tau:ligation}
\vspace*{0.2cm}
\begin{tabular}{cccccccc}
\toprule
Sample&DNA&10$\times$ Buffer&Adapter&1.5 M NaCl&T4&H$_2$O&Total\\
\midrule
1&35&4.40&1.00&1.10&2.00&0.50&44.00\\
3&35&4.40&1.00&1.10&2.00&0.50&44.00\\
5&35&4.40&1.00&1.10&2.00&0.50&44.00\\
6&35&4.40&1.00&1.10&2.00&0.50&44.00\\
\bottomrule
\end{tabular}
\end{table}

\subsection{Adapter ligation}
Following \cite{Etter2011}, I was planning to use NEB Buffer 2, supplemented with rATP, to run the ligation.  The reason argued to suggest NEB buffer 2, instead of regular ligation buffer, is that it contains 50 mM NaCl, which is about the maximum that T4 DNA ligase can stand without being inhibited. The presence of some salt helps keep the DNA fragments annealed. When I didn't have NEB Buffer 2, I just supplemented the reaction with 50 mM NaCl.
\begin{enumerate}
\item Prepare a 1.5 M solution of NaCl in a 10 ml Falcon tube, using 10 ml H$_2$O and 0.87 g NaCl (molecular weight 58.44 g/mol).
\item Add to the DNA samples the corresponding amounts of 10$\times$ T4 ligase reaction buffer, working stock of adapters, 1.5 M NaCl, and T4 DNA ligase (2000000 U/ml), according to table~\ref{tau:ligation}. Be sure to add the adapters to the reaction before the ligase, to prevent religation of the genomic DNA.
\item Incubate the reaction at 16 $^{\circ}$C overnight (or at least 2 hours).
\item Heat inactivate T4 DNA ligase for 10 min at 65 $^{\circ}$C. Allow reactions to cool slowly down (2 $^{\circ}$C per 90 s, or 0.02$^{\circ}$C/s) to room temperature.
\end{enumerate}

\section{Second digestion}
Taking advantage of a smart adapter design, suggested by Sibelle Torres Vila√ßa (IGB, Berlin), that avoids the regeneration of the cut site, a second digestion with NspI will get rid of only chimeric fragments. The restriction enzyme NspI has 100\% activity in NEB buffer 2.1, which is very similar to the T4 DNA ligase reaction buffer. However, NspI would miss the 100 $\mu$g/ml BSA, absent in the ligase reaction buffer. Because I do not want to clean up the ligase reaction before the second digestion, I will bring the reaction volume from 44 $\mu$l to 50$\mu$l, by adding 5.0 $\mu$l of 10$\times$ CutSmart buffer and 1.0 $\mu$l of water. An alternative would be to dilute the ligation products further, to minimize the effect of the ligase buffer. Finally, I set up the second digestion reactions using the total ligation volume as starting DNA volume, and following table~\ref{tau:digest}. This ends up diluting the reaction.

\section{Clean up of ligation and second digestion reactions, and size selection}
Before pooling the samples together, it is important to check that the ligation worked. Unfortunately, this involves another clean-up step, with magnetic beads (see above). If we did not clean the reaction, the excess of adapter dimers would dominate the fragment size distributions and the Bioanalyzer would not get precise measures in the size range of interest. This time, the ratio of beads to sample must be lower than 1.5:1, like 1:1, in order to remove as many adapter dimers as possible. Based on the results of the last Bioanalyzer run, the `elution' volumes must be calculated with the aim of leveling up (and optimizing) the concentrations of fragments \emph{in the desired size range} among samples.

If the size selection is done with the magnetic beads, this is the time to do it. I follow reference \cite{Bronner2009}: "First a low concentration of AMPure XP beads is added to the sample to bind larger DNA fragments. In this step the beads containing the larger fragments are discarded. More beads are then added to the supernatant, increasing the amount of PEG and NaCl, so smaller fragment sizes will be bound. Next the supernatant is discarded and the beads are washed and intermediate fragments are eluted. Depending on the concentrations of PEG and NaCl in the first and final SPRI step distinct size ranges can be generated". The SPRI Select user guide gives detailed information on the steps, but not much about the specific ratios of beads to sample that need to be used for specific fragment size ranges. A more detailed table appears on technical note \href{http://www.gqinnovationcenter.com/documents/technicalNotes/technicalNotes_GQ15.pdf}{GQ-15} from McGill University and G√©nome Q√©bec. Based on these tables, I decide to apply a left-side selection ratio of 0.645 and a right-side selection ratio of 0.600. This is supposed to center the fragment distribution at around 550 bp, and produce a range between 300 and 900 bp. The following protocol is mostly based on Machery-Nagel's \href{http://www.gqinnovationcenter.com/documents/technicalNotes/technicalNotes_GQ15.pdf}{User's Manual} of NuceloMag$^{\copyright}$ NGS clean-up and size select. Let's say that we depart from $x$ $\mu$l of DNA sample.

\begin{enumerate}
\item If using 200 $\mu$l PCR tubes, prepare at least 400 $\mu$l of 80\% ethanol per sample. If using 1.5 ml tubes, prepare at least twice as much 80\% ethanol as total DNA volume. If ethanol is 80\%, it does not need to be freshly prepared \cite{Bronner2009}.
\item Let the magnetic beads reach room temperature for 30 minutes.
\item Vortex the bead suspension thoroughly until it appears homogeneous in colour.
\item Add exactly $0.600x$ $\mu$l of the well dispersed bead suspension to each tube.
\item Add exactly $x$ $\mu$l of DNA sample to the tubes. The volume ratio of binding buffer and bead suspension to sample is 0.6. This is the lowest of the two ratios, and it is meant for `right-selection', that is, to remove too large DNA fragments, which will stay attached to the beads. The lower this ratio, the higher the upper end of the selected size range. Adjust the pipette to $1.6x$ $\mu$l and mix by pipetting up and down 10 times.
\item Incubate the tubes at room temperature for 5 min.
\item Separate the magnetic beads against the side of the wells by placing the tubes on the magnetic separator. Wait at least 5 minutes until all the beads have been attracted to the magnets and the liquid appears clear.
\item Transfer the supernatant into new tubes and discard the beads that contain the unwanted large fragments. Make sure you have $1.6x$ $\mu$l of supernatant.
\item Vortex the bead suspension again well until it appears homogeneous in colour.
\item Add $0.045x$ $\mu$l of the well dispersed bead suspension to each tube. This amount of bead suspension adds up to the $0.600x$ $\mu$l already present in the supernatant, raising the ratio of beads to sample to 0.645. This is the left-side ratio, which controls the lower bound of the size range. Adjust the pipette $1.645x$ $\mu$l and mix by pipetting up and down 10 times.
\item Incubate the separation tubes at room temperature for 5 minutes.
\item Separate the magnetic beads against the side of the tubes by placing them on the magnetic separator. Wait at least 5 minutes untill all the beads have been attracted to the magnets or the liquid appears clear.
\item Remove and discard supernatant by pipetting. Do not disturb the beads.
\item Without removing the tubes from the magnetic separator, add at least $1.645x$ $\mu$l of 80\% ethanol, to cover all the beads without disturbing the pellet.
\item Incubate for at least 30 s, and carefully remove and discard the supernatant by pipetting.
\item Add again the same amount of 80\% ethanol without disturbing the pellet, and incubate at room temperature for at least 30 s. Then carefully remove and discard the supernatant by pipetting. Remove it completely, including residual droplets.
\item Leave the tubes on the magnetic separator and incubate at room temperature for 5 to 15 minutes in order to allow the remaining traces of alcohol to evaporate. Do not overdry the beads.
\item Remove the tubes from the magnetic separator and add 10--50 $\mu$l elution buffer (it can be water), and resuspend the pellet by pipetting up and down 10 times or by shaking.
\item Incubate the tubes at room temperature for 2--5 minutes.
\item Separate the magnetic beads against the side of the wells by placing the tubes on the magnetic separator. Wait at least 5 minutes until all the beads have been attracted to the magnets or until the liquid appears clear.
\item Transfer the supernatant containing the purified DNA fragment library to a new tube.
\end{enumerate}

\section{Amplification PCR}
The protocol for Phusion High-Fidelity PCR Master Mix with HF Buffer recommends amplification primers to be at 0.5 $\mu$M in PCR reactions of either 20 or 50 $\mu$l, which means that there should be either 10 or 25 pmols of each primer per reaction. About template DNA, the protocol only says that it should be less than 250 ng. However, the double digest protocol \cite{Peterson2012} suggests 2.0 $\mu$M primers in 20 $\mu$l of final volume; that is, 40 pmols of primers. And only 20 ng of DNA, which would be about 0.06 and 0.1 pmols of fragments, depending on their average size (500 or 200). I think the latter is closer to what I need. I run 5 PCR reactions like the one shown on table~\ref{tau:PCR}. No more than 12 cycles are recommended, to limit base misincorporation and size or sample bias.

\begin{table}
 \begin{center}
  \caption{PCR programme.}\label{tau:pcrprogramme}
  \vspace*{0.3cm}
  \begin{tabular}{lccl}
   \toprule
Step&Temp.($^\circ$C)&Time (s)&\\
   \midrule
Initial denaturation&98&30&\\
Denaturation&98&\multicolumn{1}{c|}{10}&\\
Annealing&60&\multicolumn{1}{c|}{20}&10$\times$\\
Extension&72&\multicolumn{1}{c|}{15}&\\
Final extension&72&600&\\
Hold&4&$\infty$&\\
   \bottomrule
  \end{tabular}
 \end{center}
\end{table}

\begin{table}
\begin{center}
\caption{PCR reactions in 25 $\mu$l. 10 $\mu$l of 10 $\mu$M Primer Mix may be prepared with 1 $\mu$l 100 $\mu$M Primer 1 $+$ 1 $\mu$l 100 $\mu$M Primer 2 $+$ 8 $\mu$l nuclease-free water.}\label{tau:PCR}
\vspace*{0.2cm}
\begin{tabular}{lrr}
\toprule
Component&Volume ($\mu$l)&Final conc.\\
\midrule
H$_2$O&7.75&\\
5$\times$ HF buffer&5.00&1$\times$\\
10 mM dNTPs&0.50&200 $\mu$M each\\
10 $\mu$M Primer Mix&1.25&0.5 $\mu$M\\
Template DNA&10.00&\\
Phusion DNA Pol.&0.50&0.02 U/$\mu$l\\
\midrule
Total&25.00&\\
\bottomrule
\end{tabular}
\end{center}
\end{table}


\section{Clean up with magnetic beads, and quantification}
Before sequencing, another clean up is required. And quantification is mandatory, to make sure the sequencing will work.

\bibliographystyle{abbrv}
\bibliography{gbs}
\end{document}
