---
title: "DESeq2 with phyloseq"
author: "J. Ignacio Lucas Lled√≥"
date: "16/3/2020"
output: html_document
bibliography: deseq2.bib
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(phyloseq)
library(DESeq2)
library(ggplot2)
load('../2020-02-27/ps.RData')
load('../2020-02-11/taxonomy.RData')
```

Here the goal is to test for differential abundance of taxa between age groups
(early and late life points). Data are counts: number of times an amplicon variant
has been observed. The counts depend on the library size or sequencing effort, which
varies among samples. They only carry relative information. They are *compositional*
data. Comparisons among these kinds of data are not straight forward for several
reasons. First of all, data are not normally distributed. A Poisson distribution
would adequately model count data if the average was the same in all replicates.
But there is *biological variation* among replicates: the proportion of any taxon
is not expected to be exactly the same among replicates of the same class of
samples, for biological (rather than technicall or random-sampling) reasons. Such
variation among replicates causes overdispersion, with respect to the Poisson
distribution. If not taken into account, overdispersion leads to an inflated number
of false positives [@Anders2010]. A Gamma mixture of Poisson variables produces
a Negative Binomial distribution, which more adequately models sequence count data
[@Lu2005]. The main challenge to fit such models is to estimate the dispersion
parameter (for every taxon) from a limited number of replicates. The strategy
involves sharing information among features (taxa), and assuming that taxa with
a similar average abundance must have a similar level of variation among replicates.
This kind of methods are implemented in R packages for the analysis of RNA-seq
data, such as `EdgeR` and `DSeq2`. Here, we apply `DSeq2` [@Love2014] to the analysis
of microbial composition data [@McMurdie2014].

To my knowldege, neigther `DSeq2` nor `EdgeR` take into account another particular
feature of this kind of data. The compositional nature of the counts imply a
dependence structure: one taxon cannot increase its abundance without decreasing
every other taxa's abundances. This concern is gaining attention [@Quinn2019].
However, a common understanding is perceived that if compositional changes are
small, then it does not make a big difference to take such dependence into account.

# Comparison of abundances between early and late

Below, I mostly follow this [tutorial](https://joey711.github.io/phyloseq-extensions/DESeq2.html).
The test for significance of coefficients in the Negative Binomial GLM (`nbinomWaldTest`)
throws the warning that 98 coefficients do not converge. Neither increasing the number of
iterations nor removing taxa without at least 10 observations in at least 4 samples solved
the problem. Thus, I resort to ignore those taxa from the results.

```{r deseq2}
# I had set isoline as 'ordered' just to have isoline 6 in first position in plots.
# Now, I should *disorder* it.
sample_data(ps)$isoline <- factor(sample_data(ps)$isoline, ordered=FALSE)
ds <- phyloseq_to_deseq2(ps, ~ seqrun + isoline + age)

# ds <- DESeq(ds, test='Wald', fitType='parametric')
# The DESeq function performs three steps that can be run independently:
# ds <- estimateSizeFactors(ds)
stopifnot(all.equal(rownames(SeqTabNoChim), colnames(ds)))
sizeFactors(ds) <- rowSums(SeqTabNoChim)
# This filtering did not improve convergence:
#   f  <- rowSums(counts(ds) >= 10) >= 4
#   ds <- ds[f,]
#   ds <- estimateSizeFactors(ds)

ds <- estimateDispersions(ds, fitType = 'local')
ds0 <- nbinomWaldTest(ds)
dim(ds0)
ds0 <- ds0[mcols(ds0)$betaConv, ]
dim(ds0)
res <- results(ds0, cooksCutoff = FALSE)
sum(is.na(res$padj))
res[is.na(res$padj),'padj'] <- 9.9999
sigtab <- res[res$padj < 0.05, ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
dim(sigtab)
```

```{r plot}
theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Set1", ...) {
    scale_fill_brewer(palette = palname, ...)
}
# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab$Phylum = factor(as.character(sigtab$Phylum), levels=names(x))
# Genus order
x = tapply(sigtab$log2FoldChange, sigtab$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtab$Genus = factor(as.character(sigtab$Genus), levels=names(x))
ggplot(sigtab, aes(x=Genus, y=log2FoldChange, color=Phylum)) + geom_point(size=6) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
ggsave('FoldChange.png')
```

```{r morePlots, include=FALSE, eval=FALSE}
sigtab.Acetobacter <- sigtab[sigtab$Genus == 'Acetobacter' & ! is.na(sigtab$Genus), ]
ggplot(sigtab.Acetobacter, aes(x=padj, y=log2FoldChange, label=rownames(sigtab.Acetobacter))) +
  geom_point() + scale_x_log10() + geom_label()
Acetobacter.Decreased <- row.names(sigtab)[sigtab$log2FoldChange < 0 & sigtab$Genus == 'Acetobacter' & ! is.na(sigtab$Genus)]
if (length(Acetobacter.Decreased > 0)) {
  ps.Aceto.Dec <- subset_taxa(ps, colnames(otu_table(ps)) %in% Acetobacter.Decreased)
  ggplot(psmelt(transformSampleCounts(ps.Aceto.Dec, function(x) x + 1)), aes(x=age, y=Abundance)) +
    geom_violin() + facet_wrap(~OTU) + scale_y_log10()
}
```

# Testing the isoline term

Not all amplicons must respond the same way to age, sequencing batch or isoline.
I should fit progressively more complex models to amplicon abundance, retaining
in every step only the amplicons that are better modeled with the additional terms.
I depart from object `ds`, with already calculated dispersions and size factors.

```{r isolineTerm, message=FALSE, warning=FALSE}
ds3 <- nbinomLRT(ds, full = design(ds),
                 reduced = ~ seqrun + age)
res3 <- results(ds3, cooksCutoff = FALSE)
res3[is.na(res3$padj), 'padj'] <- 9.9999
AmpliconsIsolineTerm <- rownames(res3[res3$padj < 0.1,])
```

Above I checked if any amplicon's abundance can be modeled without using the isoline
term. Those would be amplicons with the same response to age and sequencing batch in
all isolines: boring ones that could be excluded from further analyses. Unfortunately
they are a majority. Probably not because they are that homogeneous, but because
dispersions are too high within isolines to notice differences among them. I will
focus on the `r length(AmpliconsIsolineTerm)` most significant amplicons.

# Testing for an interaction between age and isoline

Even though unlikely, it would be very interesting to identify amplicons the abundance
of which respond to age in an isoline-specific way. There being 24 isolines, the number
of parameters escalate more than the expected improvement in the fit. But I will try,
anyways. I will use only amplicons previously shown to require the isoline term when
modeling their abundances.

```{r LRT, warning=FALSE}
ds2 <- ds[AmpliconsIsolineTerm,]
ds2 <- nbinomLRT(ds2, full = ~ seqrun + isoline + age + isoline:age,
                 reduced = ~ seqrun + isoline + age)
dim(ds2)
ds2 <- ds2[mcols(ds2)$fullBetaConv, ]
dim(ds2)
res2 <- results(ds2, cooksCutoff = FALSE)
res2[is.na(res2$padj), 'padj'] <- 9.9999
AmpliconsInteractionTerm <- rownames(res2[res2$padj < 0.1, ])
```

Surprisingly, there are `r sum(res2$padj < 0.1)` amplicons with a significant
interaction term between isoline and age. I expect these to be the most relevant
for life-history traits variance among isolines.

# Session Information

```{r sessioninfo}
save(ds, ds2, ds3, res, res2, res3, sigtab,
     AmpliconsIsolineTerm, AmpliconsInteractionTerm, file='DESeq2.RData')
sessionInfo()
```

# References
