---
title: "DESeq2 with phyloseq"
author: "J. Ignacio Lucas Lled√≥"
date: "16/3/2020"
output: html_document
bibliography: deseq2.bib
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(phyloseq)
library(DESeq2)
library(ggplot2)
load('../2020-02-27/ps.RData')
```

Here the goal is to test for differential abundance of taxa between age groups
(early and late life points). Data are counts: number of times an amplicon variant
has been observed. The counts depend on the library size or sequencing effort, which
varies among samples. They only carry relative information. They are *compositional*
data. Comparisons among these kinds of data are not straight forward for several
reasons. First of all, data are not normally distributed. A Poisson distribution
would adequately model count data if the average was the same in all replicates.
But there is *biological variation* among replicates: the proportion of any taxon
is not expected to be exactly the same among replicates of the same class of
samples, for biological (rather than technicall or random-sampling) reasons. Such
variation among replicates causes overdispersion, with respect to the Poisson
distribution. If not taken into account, overdispersion leads to an inflated number
of false positives [@Anders2010]. A Gamma mixture of Poisson variables produces
a Negative Binomial distribution, which more adequately models sequence count data
[@Lu2005]. The main challenge to fit such models is to estimate the dispersion
parameter (for every taxon) from a limited number of replicates. The strategy
involves sharing information among features (taxa), and assuming that taxa with
a similar average abundance must have a similar level of variation among replicates.
This kind of methods are implemented in R packages for the analysis of RNA-seq
data, such as `EdgeR` and `DSeq2`. Here, we apply `DSeq2` [@Love2014] to the analysis
of microbial composition data [@McMurdie2014].

To my knowldege, neigther `DSeq2` nor `EdgeR` take into account another particular
feature of this kind of data. The compositional nature of the counts imply a
dependence structure: one taxon cannot increase its abundance without decreasing
every other taxa's abundances. This concern is gaining attention [@Quinn2019].
However, a common understanding is perceived that if compositional changes are
small, then it does not make a big difference to take such dependence into account.

Below, I mostly follow this [tutorial](https://joey711.github.io/phyloseq-extensions/DESeq2.html).
The test for significance of coefficients in the Negative Binomial GLM (`nbinomWaldTest`)
throws the warning that 98 coefficients do not converge. Neither increasing the number of
iterations nor removing taxa without at least 10 observations in at least 4 samples solved
the problem. Thus, I resort to ignore those taxa from the results.

```{r deseq2}
# I had set isoline as 'ordered' just to have isoline 6 in first position in plots.
# Now, I should *disorder* it.
sample_data(ps)$isoline <- factor(sample_data(ps)$isoline, ordered=FALSE)
ds <- phyloseq_to_deseq2(ps, ~ seqrun + isoline + age)

# ds <- DESeq(ds, test='Wald', fitType='parametric')
# The DESeq function performs three steps that can be run independently:
ds <- estimateSizeFactors(ds)
# This filtering did not improve convergence:
#   f  <- rowSums(counts(ds) >= 10) >= 4
#   ds <- ds[f,]
#   ds <- estimateSizeFactors(ds)

ds <- estimateDispersions(ds)
ds <- nbinomWaldTest(ds)
dim(ds)
ds <- ds[mcols(ds)$betaConv, ]
dim(ds)
res <- results(ds, cooksCutoff = FALSE)
sum(is.na(res$padj))
res[is.na(res$padj),'padj'] <- 9.9999
sigtab <- res[res$padj < 0.001, ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps)[rownames(sigtab), ], "matrix"))
dim(sigtab)
```

```{r plot}
theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Set1", ...) {
    scale_fill_brewer(palette = palname, ...)
}
# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab$Phylum = factor(as.character(sigtab$Phylum), levels=names(x))
# Genus order
x = tapply(sigtab$log2FoldChange, sigtab$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtab$Genus = factor(as.character(sigtab$Genus), levels=names(x))
ggplot(sigtab, aes(x=Genus, y=log2FoldChange, color=Phylum)) + geom_point(size=6) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))

```

It is unexpected to see some *Acetobacter* variants increase in abundance, while others
decrease. The following plots clarify this situation.

```{r morePlots}
sigtab.Acetobacter <- sigtab[sigtab$Genus == 'Acetobacter' & ! is.na(sigtab$Genus), ]
ggplot(sigtab.Acetobacter, aes(x=padj, y=log2FoldChange, label=rownames(sigtab.Acetobacter))) +
  geom_point() + scale_x_log10() + geom_label()
Acetobacter.Decreased <- row.names(sigtab)[sigtab$log2FoldChange < 0 & sigtab$Genus == 'Acetobacter' & ! is.na(sigtab$Genus)]
ps.Aceto.Dec <- subset_taxa(ps, colnames(otu_table(ps)) %in% Acetobacter.Decreased)
ggplot(psmelt(transformSampleCounts(ps.Aceto.Dec, function(x) x + 1)), aes(x=age, y=Abundance)) +
  geom_violin() + facet_wrap(~OTU) + scale_y_log10()
```

Even though some *Acetobacter* variants seem less abundant in late life, it is clear from these
plots that the bulk of the significantly changed variants of this genus are more abundant at
late life (positive fold change). 

```{r sessioninfo}
save(ds, res, sigtab, file='DESeq2.RData')
sessionInfo()
```

# References